### PR TITLE
Round 33: Laplace far-field penalty on volume pressure

### PR DESCRIPTION
**Pull request by morganmcg1**
# Round 33 — Laplace Far-Field Penalty on Volume Pressure

## Hypothesis

Volume-pressure has a 1.95× test gap with high-error outliers (test vol_p MAE ~28 vs val ~8). The volume_pressure target is a scalar field whose gradient should be physically consistent with a divergence-free velocity field (incompressible RANS). We can add a soft penalty that the **predicted pressure gradient obeys a known structure** — specifically, that ∇²p ≈ 0 in regions far from the body (Laplacian penalty on volume points where SDF > threshold). This is a physics-informed regularizer that should suppress test-time outliers because spurious extreme predictions in heavy-tail outlier vehicles violate Laplace's equation in the far field.

Current gaps (vs. tay SOTA PR #511):
- volume_pressure: 1.95× gap
- Test vol_p outlier contamination is the primary culprit

## Mechanism

Sample K=512 per-batch volume points where SDF > 0.5m (far-field region, outside boundary layer influence). For those points, compute ∇²p = ∂²p/∂x² + ∂²p/∂y² + ∂²p/∂z² via autograd:
1. Set `coords.requires_grad_(True)` for the K sampled far-field points
2. Compute model prediction `p = volume_head(features, coords)`
3. Compute `grad_p = torch.autograd.grad(p.sum(), coords, create_graph=True)[0]`
4. Compute second derivatives: `lap_p = sum_i grad(grad_p[:,i].sum(), coords)[:,i]` for i ∈ {x, y, z}
5. Penalty: `lambda * mean(lap_p²)`

This penalty is zero-cost when the model correctly predicts smooth far-field pressure.

## Code Changes Required

Add `--laplace-penalty-lambda <float>`, `--laplace-far-sdf-threshold <float>`, `--laplace-num-points <int>` flags:

```python
# In volume head forward, when laplace_penalty_lambda > 0:
def compute_laplace_penalty(vol_coords, sdf_values, vol_predictions, lambda_val, 
                             sdf_threshold=0.5, num_points=512):
    # Select far-field points
    far_mask = sdf_values > sdf_threshold
    if far_mask.sum() < num_points:
        return torch.tensor(0.0, device=vol_coords.device)
    
    idx = torch.where(far_mask)[0][:num_points]
    coords_far = vol_coords[idx].detach().requires_grad_(True)
    
    # Re-forward through volume head with grad-enabled coords
    # Note: need to pass coords through volume encoder again
    pred_far = volume_head(features_far, coords_far)  # [K, 1]
    
    # Compute Laplacian
    grad_p = torch.autograd.grad(pred_far.sum(), coords_far, create_graph=True)[0]  # [K, 3]
    lap_p = sum(
        torch.autograd.grad(grad_p[:, i].sum(), coords_far, create_graph=True)[0][:, i]
        for i in range(3)
    )  # [K]
    
    return lambda_val * (lap_p ** 2).mean()
```

**Important implementation note:** The second-derivative computation via autograd requires `create_graph=True`. This adds to memory — ensure batch size doesn't OOM. Start with K=256 if memory is tight and report actual K used.

Log to W&B every step:
- `train/loss_laplace_far` — the actual penalty value
- `train/mean_abs_laplacian_far` — |∇²p| diagnostic (should decrease as training progresses)
- `train/loss_total` — weighted sum

## Sweep Plan (3 arms, 4-GPU DDP)

Use `--wandb_group yi-round33-laplace-farfield` for all arms.

**Arm A — λ=0 (control, standard baseline with learnable-pe)**
```bash
torchrun --standalone --nproc_per_node=4 train.py \
  --agent alphonse \
  --wandb_group yi-round33-laplace-farfield \
  --learnable-pe \
  --optimizer lion \
  --lr 1e-4 \
  --weight-decay 5e-4 \
  --clip-grad-norm 0.5 \
  --lr-warmup-epochs 1 \
  --ema-decay 0.999 \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --model-slices 128 \
  --batch-size 4 \
  --train-surface-points 65536 \
  --eval-surface-points 65536 \
  --train-volume-points 65536 \
  --eval-volume-points 65536 \
  --laplace-penalty-lambda 0.0 \
  --epochs 50
```

**Arm B — λ=1e-3 (mild physics penalty)**
```bash
torchrun --standalone --nproc_per_node=4 train.py \
  --agent alphonse \
  --wandb_group yi-round33-laplace-farfield \
  --learnable-pe \
  --optimizer lion \
  --lr 1e-4 \
  --weight-decay 5e-4 \
  --clip-grad-norm 0.5 \
  --lr-warmup-epochs 1 \
  --ema-decay 0.999 \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --model-slices 128 \
  --batch-size 4 \
  --train-surface-points 65536 \
  --eval-surface-points 65536 \
  --train-volume-points 65536 \
  --eval-volume-points 65536 \
  --laplace-penalty-lambda 1e-3 \
  --laplace-far-sdf-threshold 0.5 \
  --laplace-num-points 512 \
  --epochs 50
```

**Arm C — λ=1e-2 (stronger physics penalty)**
```bash
torchrun --standalone --nproc_per_node=4 train.py \
  --agent alphonse \
  --wandb_group yi-round33-laplace-farfield \
  --learnable-pe \
  --optimizer lion \
  --lr 1e-4 \
  --weight-decay 5e-4 \
  --clip-grad-norm 0.5 \
  --lr-warmup-epochs 1 \
  --ema-decay 0.999 \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --model-slices 128 \
  --batch-size 4 \
  --train-surface-points 65536 \
  --eval-surface-points 65536 \
  --train-volume-points 65536 \
  --eval-volume-points 65536 \
  --laplace-penalty-lambda 1e-2 \
  --laplace-far-sdf-threshold 0.5 \
  --laplace-num-points 512 \
  --epochs 50
```

## Current Baseline (target to beat)

| Metric | yi best (PR #517) |
|--------|-------------------|
| val_abupt | **9.032%** |
| surface_pressure_rel_l2_pct | 5.702% |
| wall_shear_rel_l2_pct | 10.257% |
| wall_shear_x_rel_l2_pct | 8.520% |
| wall_shear_y_rel_l2_pct | 12.907% |
| wall_shear_z_rel_l2_pct | 12.957% |
| volume_pressure_rel_l2_pct | 5.075% |

W&B project: `wandb-applied-ai-team/senpai-v1-drivaerml`

## Win Criterion

- **Merge**: val_abupt < 9.032% AND volume_pressure component improves ≥5% relative
- **Request changes**: volume_pressure improves but val_abupt doesn't beat bar — try λ=5e-3 or increase K to 1024
- **Close**: All arms regress or Laplacian penalty diverges (check `train/loss_laplace_far` — if it grows >1.0, λ is too large)

## Key Diagnostics to Report

1. `train/loss_laplace_far` per epoch (should stay small and decrease)
2. `train/mean_abs_laplacian_far` (diagnostic for how smooth far-field predictions are)
3. `volume_pressure_rel_l2_pct` separately from val_abupt (primary mechanism signal)
4. Any OOM errors — report GPU memory usage and adjust K if needed
5. Best val_abupt per arm with epoch number
6. Test-set outlier analysis: max residual vehicles before/after penalty (secondary signal)

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Check-in on Laplace far-field penalty for volume pressure.

Targets the 1.95× volume-pressure gap. The hypothesis: ∇²p ≈ 0 in the irrotational far-field (SDF > 0.5m), so a Laplacian regularizer should suppress outlier vehicles violating that physical prior without distorting the near-field where ∇²p ≠ 0.

Memory warning — double autograd will roughly double activation memory and add 30–50% wall-clock overhead per step. If OOM at K=512 vol points:
- Drop to K=256 vol points for the penalty (separate from the data-loss vol points)
- Or use stochastic Hutchinson estimator for tr(H_p) instead of full ∇²p
Please report actual K used.

Lambda sweep (0 / 1e-3 / 1e-2):
- λ=0 is the control — should match yi baseline
- λ=1e-3 is the "physics nudge" sweet spot we expect
- λ=1e-2 is aggressive — may distort near-field if mask is leaky

Mask correctness is critical: confirm SDF > 0.5m (NOT < 0.5m) and that the mask excludes the entire car interior. Log the fraction of vol points participating in the penalty per batch.

Win gates (vs yi merge bar 9.032%):
- EP2 val_abupt < 8.81%
- Final val_abupt < 9.032%
- volume_pressure_rel_l2_pct < 5.075% (current channel value) — the targeted gap

Please post W&B run IDs and EP2 results. W&B group: yi-round33-laplace-farfield.

**Comment by morganmcg1** [comment]
ADVISOR: URGENT ESCALATION — No response in >6h. You were assigned PR #579 (Laplace far-field soft penalty on vol pressure, lambda sweep {1e-3, 1e-2, 1e-1}) and have not posted any acknowledgement, launch confirmation, or W&B run ID. If you do not post a launch confirmation with a W&B run ID within the next 60 minutes, this PR will be reassigned to another student. Please respond immediately with your launch status.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #579 because ADVISOR: Closing PR #579 — Laplace far-field penalty for volume pressure.

Student alphonse has been completely non-responsive. Two advisor messages were posted: an initial check-in at 08:12:33 UTC and an urgent escalation at 09:06:39 UTC. No acknowledgement, no launch confirmation, no W&B run ID has been posted.

This hypothesis remains valid and will be reassigned to another student. The Laplace far-field penalty (∇²p ≈ 0 in irrotational far-field at SDF > 0.5m) directly targets the 1.95× volume-pressure gap and should be pursued in a future PR..
